<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, user-scalable=no, minimum-scale=1.0, maximum-scale=1.0">
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
</head>

<body>
  <div class="text-justify" style="font-size:16px">
  <h3>Introduction</h3>
  <p>
  We performed a computational analysis on the first 70 aa at <strong>the N-t (N-terminal analysis)</strong> of the
  whole proteome and on 70 aa windows starting from all <strong>internal TMDs (Internal TMDs analysis)</strong> to
  identify <strong>potential SPC cleavage sites</strong> not associated with signal peptides (<kbd>cryptic cleavage sites</kbd>) by
  using a pre-existing a signal peptide prediction software (<a href="https://pubmed.ncbi.nlm.nih.gov/21959131/">SignalP</a>) and comparing the output of two
  networks (<strong>no-TM</strong> and <strong>TM</strong> modes). Our hits represent membrane protein with type-II oriented TMD
  predicted to be cleaved in <strong><i>no-TM network</i></strong> mode but not in <strong><i>TM network</i></strong> mode. Moreover, we rank the hits based on the number
  of disease-linked mutation.
  </p>

  <h3>Computational method</h3>
  <p>
  We used the command-line version of the <a href="https://services.healthtech.dtu.dk/service.php?SignalP-4.1">SignalP4.1 </a>program. The
  program takes a proteins sequence (FASTA formatted) of 70 amino acids length as input and predicts the
  cleavage site and reports the Y-score (combined cleavage site score) in its output. To extract
  proteins with cryptic cleavage sites, we created two types of peptide sequences by considering the first 70 amino acids of all:
  <p>
  (i) proteins that lack a canonical signal peptide (<strong>N-terminal analysis</strong>)<br>
  (ii) transmembrane domains of proteins that lack a canonical signal peptide (<strong>Internal TMDs analysis</strong>)
  </p>
  We predicted the Y-scores of both types of peptide sequences using the SignalP-TM (<strong>TM network</strong> mode) and SignalP-noTM (<strong>no-TM network</strong> mode) version (see Methods for details).
  A protein/transmembrane domain was considered to have a non-canonical cleavage site if
  its corresponding peptideâ€™s Y-score in the SignalP-TM output was less than 0.6 and in SignalP-noTM output was more than 0.5. We retrieved all
  the proteins and their associated feature information from <a href="https://pubmed.ncbi.nlm.nih.gov/14681372/">UniProt/Swiss-Prot</a>.
  We assigned mutational information provided in (a)
  <a href="https://pubmed.ncbi.nlm.nih.gov/14681372/">UniProt</a>, <a href="https://pubmed.ncbi.nlm.nih.gov/30371878/">COSMIC</a>, and <a href="https://pubmed.ncbi.nlm.nih.gov/29165669/">ClinVAR</a> databases.
  </p>

  <h3>Validation experiments</h3>
  <p>
    Validation of the hits (proteins predicted to have non-canonical cleavage sites) were then performed by ectopically expressing the WT (Wild Type) and mutant constructs in
    WT HEK293T cells or lacking the regulatory SPC subunit SPCS1. Cell lysates were then analysed
    via Western blot to detect possible SPCS1-dependent cleavage fragments (see Methods for details).
  </p>
  <h3>Contact</h3>
  <p>
    <strong>Gurdeep Singh</strong>: gurdeep.singh@bioquant.uni-heidelberg.de (<a href="russelllab.org">Russell lab</a>, Heidelberg)<br>
    <strong>Andrea Zanotti</strong>: a.zanotti@zmbh.uni-heidelberg.de (<a href="https://biochemie-med.uni-koeln.de/en/research/research-groups/lemberg-lab">Lemberg lab</a>, Heidelberg/Cologne)<br>
    <strong>Matthias Feige</strong>: matthias.feige@tum.de (<a href="https://www.department.ch.tum.de/cell/home/">CPB lab</a>, Munich)
  </p>
  </div>
</body>
</html>
